UsbIpServer/RegistryUtils.cs

```diff
@@ -5,45 +5,207 @@
+            if (guid != null)
+            {
+                Registry.LocalMachine.DeleteSubKeyTree(@$"{DevicesRegistryPath}\{guid}");
+            }
+        }
+
+        public static void StopSharingDevice(string guid)
+        {
+            Registry.LocalMachine.DeleteSubKeyTree(@$"{DevicesRegistryPath}\{guid}");
+        }
+
+        public static void StopSharingAllDevices()
+        {
+            var deviceKeyNames = Registry.LocalMachine.CreateSubKey(DevicesRegistryPath).GetSubKeyNames();
+            foreach (var keyName in deviceKeyNames)
+            {
+                StopSharingDevice(keyName);
+            }
+        }
+
+        public class PersistedDevice
+        {
+            public string Guid { get; }
+            public string BusId { get; }
+            public string Name { get; }
+            public PersistedDevice(string guid, string busid, string name)
+            {
+                Guid = guid;
+                BusId = busid;
+                Name = name;
+            }
+
+        }
+
+        public static RegistryKey? GetRegistryKey(ExportedDevice device)
+        {
+            var deviceKeyNames = Registry.LocalMachine.CreateSubKey(DevicesRegistryPath).GetSubKeyNames();
+            foreach (var keyName in deviceKeyNames)
+            {
+                var deviceKey = Registry.LocalMachine.CreateSubKey(@$"{DevicesRegistryPath}\{keyName}");
+                if (IsDeviceMatch(deviceKey, device))
+                {
+                    return deviceKey;
+                }
+            }
+
+            return null;
+        }
+
+        public static string? GetRegistryKeyName(ExportedDevice device)
+        {
+            var deviceKeyNames = Registry.LocalMachine.CreateSubKey(DevicesRegistryPath).GetSubKeyNames();
+            foreach (var keyName in deviceKeyNames)
+            {
+                var deviceKey = Registry.LocalMachine.CreateSubKey(@$"{DevicesRegistryPath}\{keyName}");
+                if (IsDeviceMatch(deviceKey, device))
+                {
+                    return keyName;
+                }
+            }
+
+            return null;
+        }
+
+
+
+        public static void SetDeviceAsAttached(ExportedDevice device)
+        {
+            var key = GetRegistryKey(device);
+            if (key != null)
             {
-                Registry.LocalMachine.DeleteSubKey($@"{DevicesRegistryPath}\{busId}");
+                key.CreateSubKey("Attached", true, RegistryOptions.Volatile);
             }
         }
 
-        public static string[] GetRegistryDevices()
+        public static void SetDeviceAsDetached(ExportedDevice device)
         {
-            return Registry.LocalMachine.OpenSubKey(DevicesRegistryPath)?.GetSubKeyNames() ?? Array.Empty<string>();
+            var keyName = GetRegistryKeyName(device);
+            if (keyName != null)
+            {
+                Registry.LocalMachine.DeleteSubKeyTree(@$"{DevicesRegistryPath}\{keyName}\Attached");
```

Comment:
This is always called at service start, also if the device is currently not attached.
You need the call signature with the second argument `false`, i.e., don't fail if the subkey does not exist to begin with.

Reply:
Good catch!